### DISCUSSION TITLE
Java Module System?

### DISCUSSION THREAD

**Discussion by hmottestad** [eclipse-rdf4j/rdf4j#4942] [Q&A]
Hi,
I am in the process of upgrading from Java 8 to Java 17, and I am running into problems with RDF4J in conjunction with the Java Module System (a.k.a jigsaw)

Since version 5 is about to come out and the MANIFEST.MF states "Build-Jdk-Spec: 11", wouldn't it be a good time to become compatible with Java Module System?

It is fairly difficult to work around the issues with RDF4J's many jar files, since jar some jars contain the same packages, which Java Module System is unhappy about.

E.g. rdf4j-queryparser-api-5.0.0-M2.jar contains org.eclipse.rdf4j.query.impl, which is also present in rdf4j-query-5.0.0-M2.jar, but there are many more (removing rdf4j-sail-base, resolved 100s of compile errors).

Cheers
Niclas

**Comment by hmottestad** [comment]
I usually use maven for my projects and I'm not getting any errors when using RDF4J in a project with Spring-boot using Java 17.

**Reply by niclash** [reply]
Java Module System has a legacy mode, where one can opt to not use the Module System to its full extent (the "unnamed" module). Then there are tools to assist in the transition, so that legacy jars shows up as Modules (I am not sure about the details).

Do you have a public project with Modules enabled (i.e. module-info.java in the project's source root, src/main/java) that I can take a look at?

My question is mostly whether RDF4J intends to make Modules (i.e. module-info.java) for the whole eco system?

**Reply by hmottestad** [reply]
No, I haven't actively used the Java Modules. We did do a lot of work to get RDF4J to compile with all its dependencies and maven plugins back when Java 9 arrived.

We don't have any plans at the moment to make active use of Java Modules in RDF4J.

Any chance you could make a small maven project to reproduce the issues that you are running into?

**Reply by niclash** [reply]
With Java Modules, we can't have the same package in multiple jar files, similar to "Sealed" packages previously. I have only found one problem, and that is `org.eclipse.rdf4j.query.impl` in both `rdf4j-query-api` and `rdf4j-queryparser-api`.

And by the looks of it, it seems that an attempt to rectify that was made, as the 2 files in `rdf4j-queryparser-api/org.eclipse.rdf4j.query.impl` are present and almost identical to two files in same module `org.eclipse.rdf4j.query.parser.impl`

So, I think the simple solution is to move and overwrite the files in o.e.r.q.impl to the o.e.r.q.parser.impl package. And that `org.eclipse.rdf4j.repository.sail.SailQuery` needs to change its import. But I am not sure that this is the only actual issue.

**Reply by hmottestad** [reply]
Probably a change that you have to wait for the next major release. We haven't started thinking about when that would be yet.
